src/merkle/mmr/full.rs

```diff
@@ -202,36 +246,52 @@ impl Mmr {
     // ============================================================================================
 
     /// Internal function used to collect the Merkle path of a value.
+    ///
+    /// The arguments are relative to the target tree. To compute the opening of the second leaf
+    /// for a tree with depth 2 in the forest `0b110`:
+    ///
+    /// - `tree_bit`: Depth of the target tree, e.g. 2 for the smallest tree.
+    /// - `relative_pos`: 0-indexed leaf position in the target tree, e.g. 1 for the second leaf.
+    /// - `index_offset`: Node count prior to the target tree, e.g. 7 for the tree of depth 3.
     fn collect_merkle_path_and_value(
         &self,
         tree_bit: u32,
         relative_pos: usize,
         index_offset: usize,
-        mut index: usize,
     ) -> (RpoDigest, Vec<RpoDigest>) {
-        // collect the Merkle path
-        let mut tree_depth = tree_bit as usize;
-        let mut path = Vec::with_capacity(tree_depth + 1);
-        while tree_depth > 0 {
-            let bit = relative_pos & tree_depth;
+        // see documentation of `leaf_to_corresponding_tree` for details
+        let tree_depth = (tree_bit + 1) as usize;
+        let mut path = Vec::with_capacity(tree_depth);
+
+        // The tree walk below goes from the root to the leaf, compute the root index to start
+        let mut forest_target = 1usize << tree_bit;
+        let mut index = nodes_in_forest(forest_target) - 1;
+
+        // Loop until the leaf is reached
+        while forest_target > 1 {
+            // Update the depth of the tree to correspond to a subtree
+            forest_target >>= 1;
+
+            // compute the indeces of the right and left subtrees based on the post-order
             let right_offset = index - 1;
-            let left_offset = right_offset - nodes_in_forest(tree_depth);
```

Comment:
This was a bug, I caught it while adding tests for the partial mmr. Added two tests to cover it. Also simplified the code a bit.